### ISSUE TITLE
BaseOutput.send is not thread-safe

### ISSUE THREAD

**Issue by adarob** [olemb/mido#49]
It appears that send (at least for the PortMidi backend) is not thread-safe. I'd like to add a lock on the send method for BaseOutput. If this sounds reasonable, let me know and I will submit a PR.

**Comment by olemb** [comment]
Thanks for your pull request! I've merged it as you've probably seen.

I didn't consider thread safety when I designed Mido. A lock is a very nice and easy way to add this.

One little change I'd make is to use use the `with` statement:

    with self._lock:
        self._send(messages.copy())    

This will ensure that the lock is released even if `self._send()` raises an exception.

I think the same lock could be used in `receive()` to make reception safe as well. (There may be some issues that I haven't considered, but I think it should work.)

    while True:
        with self._lock:
            self._receive(block=block)
            if self._messages:
                return self._messages.popleft()
            elif not block:
                return None
            elif self.closed:
                raise IOError('port closed during receive()')

       sleep()

Maybe some other things should be thread safe? Closing ports?

Let me know if you want to have a go at implementing this. If you do, could you check out my `develop` thread and make your changes there? This would make the merge a little bit cleaner.

**Comment by adarob** [comment]
Hi, sorry I was on vacation.

I just made a PR to switch to using a switch statement: https://github.com/olemb/mido/pull/51
How often do you merge your develop branch into your master? We are pulling from master in our project (https://github.com/tensorflow/magenta).

Yeah, it looks like reception is also not threadsafe. I'll have a look on Monday and try to add locking where it is needed.

**Comment by olemb** [comment]
No problem.

I merge when I release a new version, which I like to do fairly quickly after something has been fixed or added. I see that it's been over a year since the last release and that's way too long.

I'll release a new version when we've completed the locking. Thanks for offering to help out.

**Comment by olemb** [comment]
I wonder if the close() method should also be lock protected. In case someone uses it after it's closed it would be nice if it raised an exception instead of crashing.

By the way, cool project!

**Comment by olemb** [comment]
I've merged your changes. I've been a bit back and forth on the decorators but I think they are the way to go. They make the code a lot clearer.

The decorator function was only indented 2 spaces for some reason. I've fixed this.

Now all that remains is to update the docs. I'll hopefully find the time for this today.

**Comment by adarob** [comment]
Sounds good! Two spaces is how we do it internally at Google. I always
forget that it's not that way elsewhere :)

-Adam

**Comment by olemb** [comment]
No problem. :)

I ran into problems with the locking in receive(). If another thread was inside receive(block=True) a call to receive(block=False) would have to wait for the lock.

Since there is no way around this I decided to switch back to explicit locks.

By the way, I was not aware of RLock. That's really useful.

**Comment by adarob** [comment]
Yes, RLock is pretty amazing!!

It sounds like we might want a condition variable for receive. I'll have a
look.

**Comment by adarob** [comment]
So, first off I'm really sad to see that PortMidi doesn't support blocking
reads. I hadn't realized this, and it seems problematic for something like
music where timing is very important.

It looks like polling is therefore unavoidable in _receive. However, we can
get rid of the polling in receive, which would potentially halve the lag
and reduce a decent amount of wasted CPU usage. You would do this by having
a condition variable that releases the lock and is notified by the polling
thread when a message is received.  I highly recommend we do this.

This would also get rid of the deadlock issue you pointed out above.

Even if you don't want to use the condition variable, I also recommend
sticking with the decorator but keeping the receive function how you have
it now. It doesn't have to be all or nothing.

-Adam

**Comment by adarob** [comment]
Adding @cghawthorne.

**Comment by olemb** [comment]
Yes, I would love to replace the polling with something more efficient. I
don't quite understand your suggestion though. Could you show me some
example code?

**Comment by adarob** [comment]
Sure, I'll implement it and send you a PR. Might not happen today though.

**Comment by adarob** [comment]
Here is an incomplete solution for you to look at: https://github.com/adarob/mido/commit/c03745086ee568fbeb85c85cd2d7ebc0924b0b54

I can provide more details if needed, but essentially the receive thread will block until the backend polling thread returns. This will move one of the spin-locks. Unfortunately it won't be possible to remove the other without support in the backend libraries.

**Comment by olemb** [comment]
Neat. That would fix the waiting problem, but unfortunately only for RtMidi ports. Since PortMidi has no native support callbacks I had to implement them with a thread that polls and waits. Also, it would not work with non-backend ports like `MultiPort` and `SocketPort`.

I've been thinking of rewriting the RtMidi backend to use a queue internally, and having the RtMidi callback parse incoming bytes and put them in the queue. Then `receive()` can could just get messages from the queue. This should be clean and simple.

Since I'm very far behind in releasing changes and I got a request for thread safe ports today I will release the code as it is now and continue working on this for the next point release.

**Comment by olemb** [comment]
I've released 1.1.15 so now the master branch should be thread safe.

**Comment by olemb** [comment]
I've implemented the queue idea:

    https://github.com/olemb/mido/blob/rtmidi-queue/mido/backends/rtmidi.py

**Comment by adarob** [comment]
Nice!
To handle the ctrl-c problem, I think you can do something like:

```
try:
    queue.get(block=block)
except KeyboardInterrupt:
    return None
```

**Comment by olemb** [comment]
Thanks.

I'm sort of having the opposite problem though: I want people to be able to break out of the `receive()` with CTRL-C. This works out of the box in Python 3 but is more problematic in Python 2 where  `queue.get()` doesn't respond to keyboard interrupts leaving the thread hanging.

There is a way around this: you can call `queue.get()` with a timeout. Somehow this makes it respond to keyboard interrupts. (I haven't dug around in the code to find out why.) The signal still doesn't propagate to background threads though so they will still hang.

I see two options:

* use true blocking in Python 2 but warn in the docs that background threads may hang
* just use poll and wait for Python 2

I'm leaning towards the second option.

**Comment by adarob** [comment]
Indeed, this bug addressed the KeyboardInterrupt in Python 3 https://bugs.python.org/issue8844, but it has not been backported to 2. 

I recently stumbled across the reason calling get() with a timeout works, which is that the timeout implementation actually does polling, but it leads to delays in the return, potentially in the 10s of milliseconds. 

Have you pushed any of the code using queues? One potential solution is to put a "wake message" on the queue when the destructor is being called. I'm doing that in some of my code.

**Comment by olemb** [comment]
I have the queue implementation in this branch:

    https://github.com/olemb/mido/tree/rtmidi-queue

Ah, I was afraid there might be some polling when you use a timeout. 10s of
milliseconds is clearly unacceptable and much worse that the current
solution. I'm wondering if there is a way to do this at all in Python 2.

Python 3 should be good to go though, as far as I can see. (At least
versions after issue 8844 was fixed.)

**Comment by adarob** [comment]
Agreed that the current solution is better than dealing with the timing issues.

**Comment by olemb** [comment]
I used `strace` to see which system calls Python makes in the different cases.

## Python 2 and 3 with the current implementation (1.1.16 / master branch):

```
select(0, NULL, NULL, NULL, {0, 1000})  = 0 (Timeout)
select(0, NULL, NULL, NULL, {0, 1000})  = 0 (Timeout)
select(0, NULL, NULL, NULL, {0, 1000})  = 0 (Timeout)
...
```

So, 1 millisecond as returned by `mido.ports.get_sleep_time()`.


## Python 3 with `queue.get()`:

```
futex(0x2ad2710, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, ffffffff
```

This actually blocks. Clearly what we want.


## Python 2 with `queue.get()`:

```
futex(0x20647d0, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, ffffffff
```

This also blocks but doesn't respond to CTRL. Sometimes you can't even kill it normally but have to use `kill -9`, which is clearly not acceptable.


## Python 2 with `queue.get(timeout=100)` (100 seconds):

```
select(0, NULL, NULL, NULL, {0, 50000}) = 0 (Timeout)
select(0, NULL, NULL, NULL, {0, 50000}) = 0 (Timeout)
select(0, NULL, NULL, NULL, {0, 50000}) = 0 (Timeout)
```

This will cause a maximum latency of 50 milliseconds, again clearly not acceptable.

Looks like the best option is indeed `queue.get()` for Python 3 and the old poll and wait for Python 2. (I have only tested on Linux though.)

**Comment by olemb** [comment]
I've pushed a new version. Should be ready to merge now. (I just need to add a doc string to `receive()`.)

**Comment by olemb** [comment]
I've merged rtmidi-queue into my develop branch. It willl be included in 1.1.17.

**Comment by olemb** [comment]
1.1.17 is out. Thanks for your help!

**Comment by adarob** [comment]
Excellent, thanks!

On Oct 6, 2016 6:38 AM, "Ole Martin Bjørndalen" <notifications@github.com>
wrote:

1.1.17 is out. Thanks for your help!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/olemb/mido/issues/49#issuecomment-251963830>, or mute
the thread

**Comment by olemb** [comment]
Hi Adam,

I've refactored the RtMidi backend. It was getting very hard to understand with the multiple diamond inheritance and locking in the base class so I moved everything into one class: https://github.com/olemb/mido/blob/develop/mido/backends/rtmidi.py

It relies on the queue for thread safety in reception and a lock for thread safety in sending. The code may require a few tweaks but it looks very promising.
